fpsd/features.py

```diff
@@ -0,0 +1,765 @@
+
+        query = ("CREATE TABLE features.undefended_frontpage_examples AS ( "
+                 "SELECT foo.exampleid FROM ( SELECT exampleid, "
+                 "count(*) FROM raw.frontpage_traces GROUP BY exampleid) foo);")
+
+        self.execute_query_from_string(query)
+        return None
+
+    def _create_temp_packet_positions(self, outgoing_only=True):
+        """This method takes all rows in raw.frontpage_traces
+        and creates a temporary table packet_positions
+        with the following format:
+
+        exampleid  | rank     |  ingoing
+        (integer)  | (bigint) |  (boolean)
+        ------------------------------------
+        9          | 4        | f
+        9          | 5        | f
+        9          | 15       | f
+
+        where rank is the "position" in the trace for all
+        cells. For example, for the first outgoing cell in the
+        trace with exampleid=9 was the 4th packet in the trace.
+        This table is used by the position-based feature
+        generation functions.
+
+        Args:
+            outgoing_only [boolean]: describes whether or not we should
+                only use outgoing cells to create this table
+        """
+
+        self.drop_stale_feature_table("packet_positions")
+
+        if outgoing_only:
+            where_only_outgoing = "WHERE ingoing <= false"
+        else:
+            where_only_outgoing = ""
+
+        query = ("CREATE TEMP TABLE packet_positions AS    "
+                 "(SELECT exampleid, rank, ingoing                  "
+                 "  FROM (                                          "
+                 "    SELECT                                        "
+                 "      ROW_NUMBER() OVER                           "
+                 "      (PARTITION BY exampleid ORDER BY t_trace)   "
+                 "      AS rank,                                    "
+                 "      t.*                                         "
+                 "    FROM raw.frontpage_traces t)                  "
+                 "x {} );").format(where_only_outgoing)
+
+        self.execute_query_from_string(query)
+        return None
+
+    def _create_table_outgoing_cell_positions(self, num_cells=500):
+        """This method takes the first num_cells rows in
+        packet_positions and creates a temporary
+        table top_(num_cells)_outgoing_packet_positions
+        with the following format:
+
+        exampleid  | rank     |  outgoing_cell_position
+        (integer)  | (bigint) |  (bigint)
+        -----------------------------------------------
+        9          | 4        | 1
+        9          | 5        | 2
+        9          | 15       | 3
+
+        For example, the third outgoing cell was the 15th in the trace.
+        This table is used by the position-based feature
+        generation functions.
+        """
+        table_name = "top_{}_outgoing_packet_positions".format(num_cells)
```

Comment:
Ah in this case it doesn’t matter, because as it is a temporary table it gets auto dropped